service/common/src/main/java/org/apache/polaris/service/events/PolarisEventListener.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.events;
+
+/**
+ * Represents an event listener that can respond to notable moments during Polaris's execution.
+ * Users can either extend this interface and implement handlers for all events or, for ease, extend
+ * DefaultPolarisEventListener and only have to handle a subset of events. Event details are
+ * documented under the event objects themselves.
+ */
+public interface PolarisEventListener {
+  void onBeforeRequestRateLimited(BeforeRequestRateLimitedEvent event);
+
+  void onBeforeTableCommit(BeforeTableCommitEvent event);
+
+  void onAfterTableCommit(AfterTableCommitEvent event);
```

Comment:
Generally speaking, what are the guarantees about "after" commits? Are they triggered if the operation failed, or just when they succeeded? But then, do we need events for failed operations as well?

I see that `AfterAttemptTaskEvent` has a `boolean success` flag, but other "after" events do not. Also, maybe the full throwable is preferable than just a boolean flag.

Reply:
I agree that we should be as explicit as possible about this, although I think it might be ambitious to guarantee the exact semantics for each event out of the gate. At the very least, we should try our best to define the behavior with tests.

Reply:
I think we just need to rename to make it clear. I suggested to use a name like `TableCommitted` to indicate this event happens only if the table committed successfully.  https://github.com/apache/polaris/pull/922#discussion_r1984269545

Reply:
The Javadoc should describe the semantics to a reasonable extent. It does describe the table commit event as not being emitted during failure, and that each _attempt_ of a task regardless of success is emitted.

Docs aside, there's a question of which semantics we choose to actually implement, and since this is meant to be a broad interface there will be differing opinions on which make the most sense. Rather than trying to pick the perfect one for everyone, folks can always add new events and/or fields to existing events that represent different semantics.